cards/portfolio/addon/components/isolated.js

```diff
@@ -1,10 +1,80 @@
 import LiveIsolatedCard from 'portfolio-common/components/live-isolated-card';
 import layout from '../templates/isolated';
+import { inject as service } from '@ember/service';
+import { computed } from '@ember/object';
+import { task, waitForProperty } from 'ember-concurrency';
 
 export default LiveIsolatedCard.extend({
   layout,
+  store: service(),
+  web3: service(),
+  erc20: service(),
   isDismissed: false,
   displayGroup: 'all',
+  metamaskWallet: null,
+
+  isWeb3Loaded: computed.not('web3.isLoading'),
+
+  async init() {
+    this._super(arguments);
+
+    if (this.web3.provider && this.web3.provider.isMetaMask) {
+      await this.getMetamaskWallet.perform();
+    }
+  },
+
+  getMetamaskWallet: task(function * () {
+    yield waitForProperty(this, 'isWeb3Loaded');
+
+    let address = this.web3.address;
+    let metamaskWallet = yield this.store.findRecord('wallet', 'metamask-wallet');
+
+    this.set('metamaskWallet', metamaskWallet);
+
+    yield this.getAssetsForWallet.perform(address);
+
+    this.content.get('wallets').pushObject(this.get('metamaskWallet'));
+
+    yield this.content.save();
+  }),
+
+  getAssetsForWallet: task(function * (address) {
+    yield this.addAssetToWallet.perform(address, 'ether');
+
+    for (let token of this.erc20.tokens()) {
+      let tokenName = token.symbol.toLowerCase();
+      let balanceOf;
+
+      try {
+        balanceOf = yield this.store.findRecord(`${tokenName}-token-balance-of`, address.toLowerCase());
+      } catch (e) {
+        return;
+      }
+
+      if (balanceOf) {
+        yield this.addAssetToWallet.perform(`${address}_${tokenName}-token`, tokenName);
```

Comment:
this will also be painful when there are more than 100 tokens, writes are much slower than reads, so this will get very bogged down. I suggest that we get rid of wallets altogether. keeping them around increases our tech debt...

Reply:
If we wanted to keep wallets, a better approach woul dbe to add middleware that would iterate through all the tokens we knew about on the back end, so that the client makes a single request to inform the server of the address, and the server iterates through all the tokens without requiring the client to issue a request for each token.